python/ray/tune/syncer.py

```diff
@@ -142,395 +79,408 @@ class SyncConfig:
-            return self.worker_ip, self._remote_dir
-        return "{}@{}:{}/".format(ssh_user, self.worker_ip, self._remote_dir)
+    def wait(self):
+        if self._sync_process:
+            try:
+                self._sync_process.wait()
+            except Exception as e:
+                raise TuneError(f"Sync process failed: {e}") from e
+            finally:
+                self._sync_process = None
+
+    def retry(self):
+        if not self._current_cmd:
+            raise TuneError("No sync command set, cannot retry.")
+        cmd, kwargs = self._current_cmd
+        self._sync_process = _BackgroundProcess(cmd)
+        self._sync_process.start(**kwargs)
 
 
 @DeveloperAPI
-def get_cloud_syncer(
-    local_dir: str,
-    remote_dir: Optional[str] = None,
-    sync_function: Optional[Union[Callable, str]] = None,
-) -> CloudSyncer:
-    """Returns a Syncer.
+def get_node_to_storage_syncer(sync_config: SyncConfig) -> Optional[Syncer]:
+    """"""
+    if sync_config.syncer is None:
+        return None
 
-    This syncer is in charge of syncing the local_dir with upload_dir.
+    if not sync_config.upload_dir:
+        return None
 
-    If no ``remote_dir`` is provided, it will return a no-op syncer.
+    if sync_config.syncer == "auto":
+        return _DefaultSyncer(sync_period=sync_config.sync_period)
 
-    If a ``sync_function`` is provided, it will return a CloudSyncer using
-    a custom SyncClient initialized by the sync function. Otherwise it will
-    return a CloudSyncer with default templates for s3/gs/hdfs.
+    if isinstance(sync_config.syncer, Syncer):
+        return sync_config.syncer
 
-    Args:
-        local_dir: Source directory for syncing.
-        remote_dir: Target directory for syncing. If not provided, a
-            no-op Syncer is returned.
-        sync_function: Function for syncing the local_dir to
-            remote_dir. If string, then it must be a string template for
-            syncer to run. If not provided, it defaults
-            to standard S3, gsutil or HDFS sync commands.
-
-    Raises:
-        ValueError if malformed remote_dir.
-    """
-    key = (local_dir, remote_dir)
-
-    if key in _syncers:
-        return _syncers[key]
+    raise ValueError(
+        f"Unknown syncer type passed in SyncConfig: {type(sync_config.syncer)}. "
+        f"Note that custom sync functions and templates have been deprecated. "
+        f"Instead you can implement you own `Syncer` class. "
+        f"Please leave a comment on GitHub if you run into any issues with this: "
+        f"https://github.com/ray-project/ray/issues"
+    )
 
-    if not remote_dir:
-        _syncers[key] = CloudSyncer(local_dir, remote_dir, NOOP)
-        return _syncers[key]
 
-    if sync_function == "auto":
-        sync_function = None  # Auto-detect
+@DeveloperAPI
+class SyncerCallback(Callback):
+    """Callback to synchronize trial directories on a worker node with the driver."""
```

Comment:
Beef up that this is synchronous?

Reply:
I actually don't quite get why one is sync and one is async

Reply:
There is a `wait` parameter that will tell to wait for syncing to finish (or not).

Actually I guess we could remove the first `wait()` if a sync is in process and `force=False`, let me add this to the PR.